File: kaji/ts/src/providers/costs.ts
/**
 * Per-model $/1M token cost table. Updated quarterly.
 * All rates are USD per 1,000,000 tokens.
 */
export interface ModelCostEntry {
  /** Cost in USD per 1M input tokens. */
  inputPer1M: number;
  /** Cost in USD per 1M output tokens. */
  outputPer1M: number;
}

/** Standard non-batch text-token rates for the supported provider defaults. */
const COST_TABLE: Record<string, ModelCostEntry> = {
  "gpt-5.4-mini": { inputPer1M: 0.75, outputPer1M: 4.5 },
  "claude-sonnet-4-6": { inputPer1M: 3.0, outputPer1M: 15.0 },
  "gemini-2.5-flash": { inputPer1M: 0.3, outputPer1M: 2.5 },
  "gemini-3.5-flash": { inputPer1M: 1.5, outputPer1M: 9.0 },
};

const SNAPSHOT_SUFFIX = /-(?:[0-9]{4}-[0-9]{2}-[0-9]{2}|[0-9]{8}|[0-9]{3})$/;

const MAX_SAFE_TOKEN_COUNT = 9_007_199_254_740_991;
const TOKENS_PER_RATE_UNIT = 1_000_000n;
const USD_SCALE = 10_000_000_000n;
const MAX_RATE_SIGNIFICANT_DIGITS = 32;
const MAX_RATE_FRACTIONAL_DIGITS = 32;
const MAX_RATE_ABSOLUTE_EXPONENT = 32n;
const MAX_RATE_TEXT_LENGTH = 65;
const DECIMAL_RATE = /^(0|[1-9][0-9]*)(?:\.([0-9]*[1-9]))?$/;
const SCIENTIFIC_RATE = /^([1-9])(?:\.([0-9]*[1-9]))?e(-?[1-9][0-9]*)$/;

interface DecimalParts {
  coefficient: bigint;
  scale: number;
}

function assertTokenCount(name: string, value: number): void {
  if (typeof value !== "number" || !Number.isFinite(value) || !Number.isInteger(value)) {
    throw new TypeError(`${name} must be an integer`);
  }
  if (value < 0 || value > MAX_SAFE_TOKEN_COUNT) {
    throw new RangeError(`${name} must be between 0 and ${MAX_SAFE_TOKEN_COUNT}, inclusive`);
  }
}

function boundedExponent(value: string): number {
  const negative = value.startsWith("-");
  const digits = value.replace(/^[+-]/, "");
  const normalized = digits.replace(/^0+/, "") || "0";
  if (normalized.length > 2) throw new RangeError("cost rate exponent exceeds 32");
  const magnitude = BigInt(normalized);
  if (magnitude > MAX_RATE_ABSOLUTE_EXPONENT) {
    throw new RangeError("cost rate exponent exceeds 32");
  }
  const bounded = Number(magnitude);
  return negative ? -bounded : bounded;
}

function numericRateText(value: number): string {
  if (!Number.isFinite(value)) {
    throw new TypeError("cost rate must be a finite number or canonical decimal string");
  }
  if (value < 0) throw new RangeError("cost rate must be non-negative");
  if (value === 0) return "0";
  const [rawMantissa, exponentText] = value.toString().toLowerCase().split("e");
  const mantissa = rawMantissa!.includes(".")
    ? rawMantissa!.replace(/0+$/, "").replace(/\.$/, "")
    : rawMantissa!;
  if (exponentText === undefined) return mantissa;
  const exponent = boundedExponent(exponentText);
  return exponent === 0 ? mantissa : `${mantissa}e${exponent}`;
}

function decimalParts(value: string | number): DecimalParts {
  if (typeof value !== "string" && typeof value !== "number") {
    throw new TypeError("cost rate must be a finite number or canonical decimal string");
  }
  const source = typeof value === "number" ? numericRateText(value) : value;
  if (source.length === 0 || source.length > MAX_RATE_TEXT_LENGTH) {
    throw new RangeError("cost rate exceeds the canonical length bound");
  }
  let match = DECIMAL_RATE.exec(source);
  let exponent = 0;
  if (!match) {
    match = SCIENTIFIC_RATE.exec(source);
    if (!match) throw new TypeError("cost rate is not a canonical non-negative decimal");
    exponent = boundedExponent(match[3]!);
  }
  const fraction = match[2] ?? "";
  const significant = `${match[1]}${fraction}`.replace(/^0+/, "") || "0";
  if (significant.length > MAX_RATE_SIGNIFICANT_DIGITS) {
    throw new RangeError("cost rate exceeds 32 significant digits");
  }
  if (fraction.length > MAX_RATE_FRACTIONAL_DIGITS) {
    throw new RangeError("cost rate exceeds 32 fractional digits");
  }
  let coefficient = BigInt(`${match[1]}${fraction}`);
  let scale = fraction.length - exponent;
  if (scale < 0) {
    coefficient *= 10n ** BigInt(-scale);
    scale = 0;
  }
  return { coefficient, scale };
}

function canonicalUsd(units: bigint): string {
  const whole = units / USD_SCALE;
  const fraction = (units % USD_SCALE).toString().padStart(10, "0").replace(/0+$/, "");
  return fraction ? `${whole}.${fraction}` : whole.toString();
}

/** @internal Shared-fixture entry point; not exported from the package root. */
export function calculateCostFromRatesUsdCanonical(
  inputTokens: number,
  outputTokens: number,
  inputPer1M: string | number,
  outputPer1M: string | number,
): string {
  assertTokenCount("inputTokens", inputTokens);
  assertTokenCount("outputTokens", outputTokens);
  const input = decimalParts(inputPer1M);
  const output = decimalParts(outputPer1M);
  const rateScale = Math.max(input.scale, output.scale);
  const numerator =
    BigInt(inputTokens) * input.coefficient * 10n ** BigInt(rateScale - input.scale) +
    BigInt(outputTokens) * output.coefficient * 10n ** BigInt(rateScale - output.scale);
  const denominator = TOKENS_PER_RATE_UNIT * 10n ** BigInt(rateScale);
  const scaled = numerator * USD_SCALE;
  let units = scaled / denominator;
  const remainder = scaled % denominator;
  const comparison = remainder * 2n - denominator;
  if (comparison > 0n || (comparison === 0n && units % 2n === 1n)) units += 1n;
  return canonicalUsd(units);
}

/**
 * Look up cost for an exact supported model or a provider snapshot suffix.
 * Returns `undefined` rather than guessing for unknown or routed models.
 */
export function lookupCost(model: string): ModelCostEntry | undefined {
  if (model in COST_TABLE) return COST_TABLE[model];
  const base = model.replace(SNAPSHOT_SUFFIX, "");
  return base === model ? undefined : COST_TABLE[base];
}

/** @internal Shared-fixture entry point; not exported from the package root. */
export function calculateCostUsdCanonical(
  model: string,
  inputTokens: number,
  outputTokens: number,
): string {
  assertTokenCount("inputTokens", inputTokens);
  assertTokenCount("outputTokens", outputTokens);
  const entry = lookupCost(model);
  if (!entry) return "0";
  return calculateCostFromRatesUsdCanonical(
    inputTokens,
    outputTokens,
    entry.inputPer1M,
    entry.outputPer1M,
  );
}

/**
 * Calculate cost in USD for a given token count and model.
 * Returns 0 if the model is not in the cost table.
 */
export function calculateCostUsd(model: string, inputTokens: number, outputTokens: number): number {
  return Number(calculateCostUsdCanonical(model, inputTokens, outputTokens));
}

File: kaji/packages/py/src/runtime/providers/costs.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, ROUND_HALF_EVEN, localcontext
import math
import re
from typing import Optional


@dataclass(frozen=True)
class CostEntry:
    input_per_1m: float
    output_per_1m: float


COST_TABLE: dict[str, CostEntry] = {
    "gpt-5.4-mini": CostEntry(input_per_1m=0.75, output_per_1m=4.5),
    "claude-sonnet-4-6": CostEntry(input_per_1m=3.0, output_per_1m=15.0),
    "gemini-2.5-flash": CostEntry(input_per_1m=0.3, output_per_1m=2.5),
    "gemini-3.5-flash": CostEntry(input_per_1m=1.5, output_per_1m=9.0),
}

_MAX_SAFE_INTEGER = 9_007_199_254_740_991
_TOKENS_PER_RATE_UNIT = Decimal(1_000_000)
_USD_QUANTUM = Decimal("0.0000000001")
_MAX_RATE_SIGNIFICANT_DIGITS = 32
_MAX_RATE_FRACTIONAL_DIGITS = 32
_MAX_RATE_ABSOLUTE_EXPONENT = 32
_MAX_RATE_TEXT_LENGTH = 65
_MAX_RATE_INTEGER = 10**_MAX_RATE_SIGNIFICANT_DIGITS - 1
_DECIMAL_RATE = re.compile(r"^(0|[1-9][0-9]*)(?:\.([0-9]*[1-9]))?$")
_SCIENTIFIC_RATE = re.compile(r"^([1-9])(?:\.([0-9]*[1-9]))?e(-?[1-9][0-9]*)$")
_SNAPSHOT_SUFFIX = re.compile(r"-(?:[0-9]{4}-[0-9]{2}-[0-9]{2}|[0-9]{8}|[0-9]{3})$")


def lookup_cost(model: str) -> Optional[CostEntry]:
    exact = COST_TABLE.get(model)
    if exact is not None:
        return exact
    base = _SNAPSHOT_SUFFIX.sub("", model)
    return COST_TABLE.get(base) if base != model else None


def _validate_token_count(name: str, value: int) -> None:
    if type(value) is not int:
        raise TypeError(f"{name} must be an integer")
    if not 0 <= value <= _MAX_SAFE_INTEGER:
        raise ValueError(f"{name} must be between 0 and {_MAX_SAFE_INTEGER}, inclusive")


def _bounded_exponent(value: str) -> int:
    digits = value.removeprefix("+").removeprefix("-")
    normalized = digits.lstrip("0") or "0"
    if len(normalized) > 2:
        raise ValueError("cost rate exponent exceeds 32")
    magnitude = int(normalized)
    if magnitude > _MAX_RATE_ABSOLUTE_EXPONENT:
        raise ValueError("cost rate exponent exceeds 32")
    return -magnitude if value.startswith("-") else magnitude


def _numeric_rate_text(value: int | float) -> str:
    if type(value) not in (int, float) or (
        isinstance(value, float) and not math.isfinite(value)
    ):
        raise TypeError("cost rate must be a finite number or canonical decimal string")
    if value < 0:
        raise ValueError("cost rate must be non-negative")
    if type(value) is int and value > _MAX_RATE_INTEGER:
        raise ValueError("cost rate exceeds 32 significant digits")
    if value == 0:
        return "0"
    mantissa, separator, exponent_text = str(value).lower().partition("e")
    if "." in mantissa:
        mantissa = mantissa.rstrip("0").rstrip(".")
    if not separator:
        return mantissa
    exponent = _bounded_exponent(exponent_text)
    if exponent == 0:
        return mantissa
    return f"{mantissa}e{exponent}"


def _rate_decimal(value: str | int | float) -> Decimal:
    text = value if isinstance(value, str) else _numeric_rate_text(value)
    if not 0 < len(text) <= _MAX_RATE_TEXT_LENGTH:
        raise ValueError("cost rate exceeds the canonical length bound")
    match = _DECIMAL_RATE.fullmatch(text)
    exponent = 0
    if match is None:
        match = _SCIENTIFIC_RATE.fullmatch(text)
        if match is None:
            raise ValueError("cost rate is not a canonical non-negative decimal")
        exponent = _bounded_exponent(match.group(3))
    fraction = match.group(2) or ""
    significant = (match.group(1) + fraction).lstrip("0") or "0"
    if len(significant) > _MAX_RATE_SIGNIFICANT_DIGITS:
        raise ValueError("cost rate exceeds 32 significant digits")
    if len(fraction) > _MAX_RATE_FRACTIONAL_DIGITS:
        raise ValueError("cost rate exceeds 32 fractional digits")
    if abs(exponent) > _MAX_RATE_ABSOLUTE_EXPONENT:
        raise ValueError("cost rate exponent exceeds 32")
    return Decimal(text)


def _calculate_cost_from_rates_usd_canonical(
    input_tokens: int,
    output_tokens: int,
    input_per_1m: str | int | float,
    output_per_1m: str | int | float,
) -> str:
    _validate_token_count("input_tokens", input_tokens)
    _validate_token_count("output_tokens", output_tokens)
    with localcontext() as context:
        context.prec = 128
        total = (
            Decimal(input_tokens) * _rate_decimal(input_per_1m)
            + Decimal(output_tokens) * _rate_decimal(output_per_1m)
        ) / _TOKENS_PER_RATE_UNIT
        rounded = total.quantize(_USD_QUANTUM, rounding=ROUND_HALF_EVEN)
    return format(rounded, "f").rstrip("0").rstrip(".") or "0"


def _calculate_cost_usd_canonical(
    model: str, input_tokens: int, output_tokens: int
) -> str:
    _validate_token_count("input_tokens", input_tokens)
    _validate_token_count("output_tokens", output_tokens)
    entry = lookup_cost(model)
    if entry is None:
        return "0"
    return _calculate_cost_from_rates_usd_canonical(
        input_tokens,
        output_tokens,
        entry.input_per_1m,
        entry.output_per_1m,
    )


def calculate_cost_usd(model: str, input_tokens: int, output_tokens: int) -> float:
    return float(_calculate_cost_usd_canonical(model, input_tokens, output_tokens))
